optuna/trial/_trial.py

```diff
@@ -63,10 +65,40 @@ def _init_relative_params(self) -> None:
 
         study = pruners._filter_study(self.study, trial)
 
-        self.relative_search_space = self.study.sampler.infer_relative_search_space(study, trial)
-        self.relative_params = self.study.sampler.sample_relative(
-            study, trial, self.relative_search_space
-        )
+        trial_system_attrs = self.system_attrs
+        if "relative_search_space" in trial_system_attrs:
+            self._load_relative_params(study, trial, trial_system_attrs)
+        else:
+            self.relative_search_space = self.study.sampler.infer_relative_search_space(
+                study, trial
+            )
+            self.relative_params = self.study.sampler.sample_relative(
+                study, trial, self.relative_search_space
+            )
+            self._write_relative_search_space_and_params()
+
+    def _write_relative_search_space_and_params(self) -> None:
+        relative_search_space_dict = {
+            k: distribution_to_json(d) for k, d in self.relative_search_space.items()
+        }
+        self.set_system_attr("relative_search_space", relative_search_space_dict)
+        self.set_system_attr("relative_params", self.relative_params)
+
+    def _load_relative_params(
```

Comment:
How about renaming the method name like `_read_relative_search_space_and_params` to make it consistent with `_write_relative_search_space_and_params`?

Reply:
It makes sense. Revised in bea69bb.